### PR TITLE
Add simple workflow executor when custom workflow executor is missing.

### PR DESCRIPTION
**Pull request by hasoxy**
**Issue link:** *required*
https://github.com/wso2/carbon-apimgt/issues/8451

**Support Ticket** 

**Applicable Labels:**
200
------

### Development Checklist

1. [X] Built complete solution with pull request in place.
2. [X] Ran checkstyle plugin with pull request in place.
3. [X] Ran Findbugs plugin with pull request in place.
4. [ ] Ran FindSecurityBugs plugin and verified report.
5. [X] Formatted code according to WSO2 code style.
6. [X] Have you verify the PR does't commit any keys, passwords, tokens, usernames, or other secrets?
7. [ ] Migration scripts written (if applicable).
8. [X] Have you followed secure coding standards in [WSO2 Secure Engineering Guidelines](http://wso2.com/technical-reports/wso2-secure-engineering-guidelines)? 

### Testing Checklist

1. [ ] Written unit tests.
2. [ ] Documented test scenarios(link available in guides).
3. [ ] Written automation tests (link available in guides).
4. [ ] Verified tests in multiple database environments (if applicable).
5. [ ] Verified tests in multiple deployed specifications (if applicable).
6. [ ] Tested with OBBI enabled  (if applicable).
7. [ ] Tested with specification regulatory conformance suites  (if applicable).

**Automation Test Details**

| Test Suite        | Test Script IDs   |
| ----------------- | ----------------- |
| Integration Suite | *TCXXXXX, TCXXXX* |

**Conformance Tests Details**

| Test Suite Name  | Test Suite Version | Scenarios  | Result   |
| ---------------- | ------------------ | ---------- | -------- |
| *Security Suite* | *VX.X*             | *Foo, Bar* | *Passed* |

## Resources

**Knowledge Base:** https://sites.google.com/wso2.com/open-banking/

**Guides:** https://sites.google.com/wso2.com/open-banking/developer-guides

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/workflow/TenantWorkflowConfigHolder.java [hunk 1]
```diff
@@ -159,7 +159,14 @@ public void load() throws WorkflowException, RegistryException {
 
             workflowElem = workflowExtensionsElem.getFirstChildWithName(
                     new QName(WorkflowConstants.APPLICATION_DELETION));
-            executorClass = workflowElem.getAttributeValue(new QName(WorkflowConstants.EXECUTOR));
+            if (workflowElem == null) {
+                workflowElem = OMAbstractFactory.getOMFactory()
+                        .createOMElement(new QName(WorkflowConstants.API_STATE_CHANGE));
+                executorClass = WorkflowConstants.DEFAULT_EXECUTOR_API_STATE_CHANGE;
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/workflow/TenantWorkflowConfigHolder.java:165]: Why are we setting default API state change executor as the default one when APPLICATION_DELETION workflow is missing?

**Review by chamilaadhi — CHANGES_REQUESTED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/workflow/TenantWorkflowConfigHolder.java [hunk 2]
```diff
@@ -152,15 +188,29 @@ public void load() throws WorkflowException, RegistryException {
...
             executorClass = workflowElem.getAttributeValue(new QName(WorkflowConstants.EXECUTOR));
-            clazz = TenantWorkflowConfigHolder.class.getClassLoader().loadClass(executorClass);
+            try {
+                clazz = TenantWorkflowConfigHolder.class.getClassLoader().loadClass(executorClass);
+            } catch (ClassNotFoundException e) {
+                clazz = TenantWorkflowConfigHolder.class.getClassLoader().loadClass(WorkflowConstants.
+                        DEFAULT_SUBSCRIPTION_DELETION);
+                workflowElem = AXIOMUtil.stringToOM(WorkflowConstants.
+                        DEFAULT_SUBSCRIPTION_DELETION_WORKFLOW_ELEMENT);
+            }
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/workflow/TenantWorkflowConfigHolder.java:198]: You could do this much simpler way. See below sample.

        try {
            clazz = TenantWorkflowConfigHolder.class.getClassLoader().loadClass(executorClass);
            workFlowExecutor = (WorkflowExecutor) clazz.newInstance();
            loadProperties(workflowElem, workFlowExecutor);
        } catch (ClassNotFoundException e) {
            workFlowExecutor = new SubscriptionCreationSimpleWorkflowExecutor();
        }
        workflowExecutorMap.put(WorkflowConstants.WF_TYPE_AM_SUBSCRIPTION_CREATION, workFlowExecutor);


As you can see, you don't have to keep constants for the default one. 

Another this is, `loadProperties(workflowElem, workFlowExecutor);` is only needed if there are additional properties defined in the workflow configuration section. Default executor does not have any default configurations. so we could `loadProperties()` in to the try block. This prevents the need to have constants for  XML element

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/test/java/org/wso2/carbon/apimgt/impl/workflow/TenantWorkflowConfigHolderTest.java [hunk 3]
```diff
@@ -136,26 +136,27 @@ public void testFailureToLoadTenantWFConfigWhenErrorWhileLoadingRegistryResource
...
-            Assert.fail("Expected WorkflowException has not been thrown when workflow executor class not found");
-        } catch (WorkflowException e) {
-            Assert.assertEquals(e.getMessage(), "Unable to find class");
-        }
-    }
-
+//    TODO: Modify the below test to check for the availability of the simple workflow when the customized
+//     workflow is not found.
+//    @Test
+//    public void testFailureToLoadTenantWFConfigWhenWFExecutorClassNotFound() throws Exception {
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/test/java/org/wso2/carbon/apimgt/impl/workflow/TenantWorkflowConfigHolderTest.java:142]: Don't skip tests

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/workflow/TenantWorkflowConfigHolder.java (hunk 2 above)
Comment (hasoxy) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/workflow/TenantWorkflowConfigHolder.java:183]: Refactored

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/test/java/org/wso2/carbon/apimgt/impl/workflow/TenantWorkflowConfigHolderTest.java (hunk 3 above)
Comment (hasoxy) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/test/java/org/wso2/carbon/apimgt/impl/workflow/TenantWorkflowConfigHolderTest.java]: Updated the test to suite the implementation

**Review by chamilaadhi — APPROVED** [review]

**Comment by chamilaadhi** [comment]
New PR https://github.com/wso2/carbon-apimgt/pull/11691

**Pull request closed without merging**